File: Nim/centaur.nim
#[
    Nim lang implementation for the recursive centaur
]#

import strutils

proc `*`(a: int, b: string): string =
    var output: string = ""
    for i in 0 ..< a:
        output.add(b)
    return output

proc centaur(level: int, finish: bool = true): void =
    if level < 0:
        return

    # base strings
    var
        top: string = """|       #"""
        bottom: string = ""

    # call upper levels
    centaur(level-1, false)

    if level == 0: # is head of the centaur
        top = """ o""" & '\n' & """/|\_____""" & '\n' & """|       #"""
    else: # is recursive centaur's body
        top = (level-1) * """     """ & """! !``""" & top
    
    if finish: # end of recursive centaur
        bottom = '\n' & (level) * """     """ & """! !```! !"""
    else: # recursive centaur's back
        top &= """____"""

    # print centaur lvl a
    echo top & bottom

var levels: int

let input = readLine(stdin)
levels = parseInt(input)

centaur(levels)

File: Lua/centaur.lua
TRUE = 1
FALSE = 0

-- Recursive Centaur explanation:
-- https://github.com/OakBranches/Recursive-Centaur#readme
local function centaur(level, finish)
    -- Base case: when level is less than zero, exit recursion
    if level < 0 then return end
    
    -- Recursive call for the next level
    centaur(level - 1, FALSE)

    -- Print head
    if level == 0 then
        io.write(" o\n")
        io.write("/|\\_____\n")
        io.write("|       #")
    else
        -- Print spaces for body
        for i = 1, level - 1 do
            io.write("     ")
        end
        io.write("! !  ")

        -- Print default top value
        io.write("|       #")
    end

    -- Check if it's the last level for legs
    if finish == FALSE then
        -- Print centaur's back
        io.write("____")
        io.write("\n")
    else
        io.write("\n")
        -- Print paddings
        for i = 1, level do
            io.write("     ")
        end

        -- Print legs
        io.write("! !   ! !")
    end
end

local function call_centaur()
    print("Enter level of centaur recursion: ")
    local level = tonumber(io.read())
    centaur(level, TRUE)
    return
end

call_centaur()
